File: frontend/src/app/features/processing-jobs/services/processing-job.service.ts
import { HttpClient } from '@angular/common/http'
import { inject, Injectable } from '@angular/core'
import { Observable } from 'rxjs'

import { environment } from '../../../../environments/environments.dev'
import { ApiResponse, unwrapApiResponse } from '../../../shared/utils/api-response'
import { ProcessingJobResponse, ProcessingJobStatusHistoryResponse } from '../models/processing-job.model'

@Injectable({
  providedIn: 'root'
})
export class ProcessingJobService {
  private readonly httpClient = inject(HttpClient)
  private readonly endpointUrl = `${environment.apiBaseUrl}/api/processing-jobs`

  listProcessingJobs (): Observable<ProcessingJobResponse[]> {
    return this.httpClient
      .get<ApiResponse<ProcessingJobResponse[]>>(this.endpointUrl)
      .pipe(unwrapApiResponse())
  }

  listProcessingJobsByUsername (username: string): Observable<ProcessingJobResponse[]> {
    return this.httpClient
      .get<ApiResponse<ProcessingJobResponse[]>>(this.endpointUrl, {
        params: { username }
      })
      .pipe(unwrapApiResponse())
  }

  getProcessingJob (jobId: string): Observable<ProcessingJobResponse> {
    return this.httpClient
      .get<ApiResponse<ProcessingJobResponse>>(`${this.endpointUrl}/${jobId}`)
      .pipe(unwrapApiResponse())
  }

  getProcessingJobHistory (jobId: string): Observable<ProcessingJobStatusHistoryResponse[]> {
    return this.httpClient
      .get<ApiResponse<ProcessingJobStatusHistoryResponse[]>>(
        `${this.endpointUrl}/${jobId}/history`
      )
      .pipe(unwrapApiResponse())
  }

  startProcessing (jobId: string): Observable<ProcessingJobResponse> {
    return this.httpClient
      .post<ApiResponse<ProcessingJobResponse>>(`${this.endpointUrl}/${jobId}/start`, {})
      .pipe(unwrapApiResponse())
  }

  approve (jobId: string): Observable<ProcessingJobResponse> {
    return this.httpClient
      .post<ApiResponse<ProcessingJobResponse>>(`${this.endpointUrl}/${jobId}/approve`, {})
      .pipe(unwrapApiResponse())
  }

  reject (jobId: string, reason: string): Observable<ProcessingJobResponse> {
    return this.httpClient
      .post<ApiResponse<ProcessingJobResponse>>(`${this.endpointUrl}/${jobId}/reject`, { reason })
      .pipe(unwrapApiResponse())
  }

  revoke (jobId: string, reason: string): Observable<ProcessingJobResponse> {
    return this.httpClient
      .post<ApiResponse<ProcessingJobResponse>>(`${this.endpointUrl}/${jobId}/revoke`, { reason })
      .pipe(unwrapApiResponse())
  }

}

File: backend/src/main/java/dev/lgbonillar/regreporting/processing/application/ProcessingJobService.java
package dev.lgbonillar.regreporting.processing.application;

import dev.lgbonillar.regreporting.processing.dto.ProcessingJobResponse;
import dev.lgbonillar.regreporting.processing.dto.ProcessingJobStatusHistoryResponse;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
public class ProcessingJobService {

    private final ProcessingJobQueryService processingJobQueryService;
    private final ProcessingJobWorkflowService processingJobWorkflowService;
    private final ProcessingJobHistoryService processingJobHistoryService;

    public ProcessingJobService(
            ProcessingJobQueryService processingJobQueryService,
            ProcessingJobWorkflowService processingJobWorkflowService,
            ProcessingJobHistoryService processingJobHistoryService
    ) {
        this.processingJobQueryService = processingJobQueryService;
        this.processingJobWorkflowService = processingJobWorkflowService;
        this.processingJobHistoryService = processingJobHistoryService;
    }

    public List<ProcessingJobResponse> listProcessingJobs(String username) {
        return processingJobQueryService.listProcessingJobs(username);
    }

    public ProcessingJobResponse getProcessingJob(UUID jobId) {
        return processingJobQueryService.getProcessingJob(jobId);
    }

    public ProcessingJobResponse startProcessing(UUID jobId) {
        return processingJobWorkflowService.startProcessing(jobId);
    }

    public ProcessingJobResponse completeProcessing(UUID jobId) {
        return processingJobWorkflowService.completeProcessing(jobId);
    }

    public ProcessingJobResponse failProcessing(UUID jobId, String reason) {
        return processingJobWorkflowService.failProcessing(jobId, reason);
    }

    public ProcessingJobResponse approve(UUID jobId) {
        return processingJobWorkflowService.approve(jobId);
    }

    public ProcessingJobResponse reject(UUID jobId, String reason) {
        return processingJobWorkflowService.reject(jobId, reason);
    }

    public ProcessingJobResponse revoke(UUID jobId, String reason) {
        return processingJobWorkflowService.revoke(jobId, reason);
    }

    public List<ProcessingJobStatusHistoryResponse> getProcessingJobHistory(UUID jobId) {
        return processingJobHistoryService.getProcessingJobHistory(jobId);
    }

}
